### PR TITLE
Backup missing pub/static.php file #13298

### LINKED ISSUE #13298: Backup missing pub/static.php file
When restoring from backup created using `bin/magento setup:backup --code`
following files are being ignored / missing:
```
pub/static.php
pub/static/.htpasswd
```
This is apparently a desired behaviour for backup task "excluding var and pub/static directories" via http://devdocs.magento.com/guides/v2.2/install-gde/install/cli/install-cli-backup.html

It seems though that above mentioned files files are crucial for the static content to work correctly when in default or developer mode. These missing files are never generated during static content deployment so they should be included in the backup, similar to other files from `pub/` directory.

### Preconditions
1. Magento CE 2.2.2
2. PHP-FPM 7.0.27
3. MySQL 5.7.20
4. Nginx 1.12.1

### Steps to reproduce
1. Create archive using `bin/magento setup:backup --code`
2. Untar archive into a third party location
3. Verify file list differences between
```
/tmp/original/pub/:
cron.php
errors
get.php
health_check.php
index.php
media
opt
static.php

/tmp/restored/pub/:
cron.php
errors
get.php
health_check.php
index.php
media
opt
```

following files are missing:
```
pub/static.php
```
### Expected result
Following files needed in the backup:
```
cron.php
errors
get.php
health_check.php
index.php
media
opt
static.php
```
Stop ignoring following files when creating backup:
```
pub/static.php
```

### Actual result
Following files got actually included in the backup:
```
cron.php
errors
get.php
health_check.php
index.php
media
opt
```
Thi is easy to reproduce on cli

### PR DESCRIPTION
**Pull request by ivanko-dev**
### Description (*)
Filtering files for making backup code bundle implemented with `FilterIterator` in `\Magento\Framework\Backup\Filesystem\Iterator\Filter`. This filter define which files should be skipped, comparing paths with using `strpos`,  `\Magento\Framework\Backup\Filesystem\Iterator\Filter:53`. There is rule to skip `pub/static` folder and in this case `pub/static.php` was skipped too.
According to this update filter will skip files/folders when path is completely the same or if path defined in filtering rule is parent folder.

### Fixed Issues (if relevant)

1. Fixes magento/magento2#13298 : Backup missing pub/static.php file

### Manual testing scenarios (*)

1. Run CLI command `bin/magento setup:backup --code`
2. Check compressed file generated in `var/backups`
3. File should contain `pub/static.php`

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/Backup/Filesystem/Iterator/Filter.php [hunk 1]
```diff
@@ -3,35 +3,34 @@
...
-
 class Filter extends \FilterIterator
 {
     /**
      * Array that is used for filtering
      *
      * @var array
      */
-    protected $_filters;
+    protected $filters;
```
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/Backup/Filesystem/Iterator/Filter.php:22]: Please revert this change as it's not backward compatible. More info https://devdocs.magento.com/contributor-guide/backward-compatible-development/

File: lib/internal/Magento/Framework/Setup/BackupRollback.php [hunk 2]
```diff
@@ -23,7 +23,8 @@
 class BackupRollback
 {
     /**
-     * Default backup directory
+     *
+     * @const string Default backup directory
      */
```
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/Setup/BackupRollback.php:28]: Could you just delete this phpdoc block at all? it's not adding any value.

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @ivan-koliadynskyy,

In general you changes looks good to me.
Could you review and my comments and cover your changes with some kind of automated tests?

**Comment by ihor-sviziev** [comment]
Please also notice that this functionality is deprecated: https://docs.magento.com/user-guide/system/backups.html

**Comment by ihor-sviziev** [comment]
Hi @ivan-koliadynskyy,
Will you be able to update your PR with requested changes?

**Comment by sidolov** [comment]
@ivan-koliadynskyy  I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

**Pull request closed without merging**